hw/ip_templates/otp_ctrl/dv/env/seq_lib/otp_ctrl_zeroization_vseq.sv.tpl

```diff
@@ -0,0 +1,276 @@
+    // enable_trigger_checks dist {0 :/ 4, 1 :/ 6};
+    enable_trigger_checks == 0;
+  }
+
+  constraint intr_enable_c {
+    // intr_operation_done prediction in the model needs to be revisited
+    // for now disabling operation_done checks with zeroization and enabling only otp_error
+    en_intr == 2'b10;
+  }
+
+  constraint num_trans_c {
+    num_trans  == 1;
+    num_dai_op inside {[1:10]};
+  }
+
+  constraint partition_index_c {
+    part_idx inside {
+% for part in parts_without_lc:
+      ${Name.to_camel_case(part["name"])}Idx${"" if loop.last else ","}
+% endfor
+    };
+  }
+
+  constraint dai_wr_legal_addr_c {
+% for part in parts_without_lc:
+    if (part_idx == ${Name.to_camel_case(part["name"])}Idx)
+        dai_addr inside `PART_CONTENT_RANGE(${Name.to_camel_case(part["name"])}Idx);
+% endfor
+    solve part_idx before dai_addr;
+  }
+
+
+  task body();
+    bit [TL_DW-1:0] reg_rd_val, rdata0, rdata1, rdata0_pre, rdata1_pre;
+    bit [TL_DW-1:0] addr;
+
+    bit zeroized_partition [0:LifeCycleIdx-1];
+    bit used_zeroized_addrs[bit [TL_DW-1:0]];
+
+    bit dai_regwen_state = 1;
+
+    for (int i = 0; i < NumPart; i++) begin
+      // part_idx = otp_ctrl_part_pkg::part_idx_e'(i);
+      part_idx = i;
+      `uvm_info(`gfn, $sformatf("part_idx :%s, addr offset :%x",
+                                 part_idx.name, PartInfo[i].offset), UVM_LOW);
+    end
+
+    // Adding a mechanism to ignore dai_idle checks in main smoke part of the test.
+    ignore_dai_idle_check = 1;
+
+    super.body();
+    isr_present           = 1;
+
+
+    this.rand_mode(0);
+    this.dai_addr.rand_mode(1);
+    this.part_idx.rand_mode(1);
+
+    dut_init_completed = 0;
+
+
+    fork
+      intr_service_routine();
+      check_trigger_routine();
+    join_none
+    #0 // Force the fork processes to have started
+
+    for (int i = 0; i < num_dai_op; i++) begin
+      // Now randomize the partitions and address to zeroize
+      `DV_CHECK_RANDOMIZE_FATAL(this)
+
+      // First check if the partition zeroizable field is zeroized. If not set the field
+      if (!part_is_zeroizable(part_idx)) begin
+        `uvm_info(`gfn, $sformatf("ZRing Non Zeroizable Partition: part_idx :%s, dai_addr :%x",
+                part_idx.name, dai_addr), UVM_LOW);
+        dai_zeroize(.addr(dai_addr));
+      end else begin
+        if (!zeroized_partition[part_idx]) begin
+          `uvm_info(`gfn, $sformatf("ZRing Partition: part_idx :%s, dai_addr :%x",
+                  part_idx.name, zeroized_offset(part_idx)), UVM_LOW);
+          dai_zeroize(.addr(zeroized_offset(part_idx)));
+          zeroized_partition[part_idx] = 1;
+        end
+
+        if (used_zeroized_addrs.exists(dai_addr)) begin
+          `uvm_info(`gfn, $sformatf("addr %0h is already written!", addr), UVM_MEDIUM)
+          continue;
+        end
+
+        if (set_dai_regwen && dai_regwen_state) begin
+          csr_rd(ral.direct_access_regwen, reg_rd_val);
+          cfg.clk_rst_vif.wait_clks(5);
```

Comment:
Please avoid explicit waits and use dai_idle instead: we don't expect this read to fail. In fact, try to use the dai_idle check instead of waits after any dai op.